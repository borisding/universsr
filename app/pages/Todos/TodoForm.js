import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { TextInput } from '@app/common/components';
import './TodoForm.module.scss';

const initialState = {
  isInvalid: false,
  todoInput: ''
};

export default function TodoForm(props) {
  const [isInvalid, setIsInvalid] = useState(initialState.isInvalid);
  const [todoInput, setTodoInput] = useState(initialState.todoInput);

  const styles = classNames('todo-input', { 'error-input': !!isInvalid });
  const placeholder = isInvalid ? 'Todo is required!' : 'Enter new todo...';

  function handleChange(evt) {
    setIsInvalid(initialState.isInvalid);
    setTodoInput(evt.target.value);
  }

  function handleBlur(evt) {
    if (isInvalid) {
      setIsInvalid(!isInvalid);
      setTodoInput(evt.target.value);
    }
  }

  function handleSubmit(evt) {
    evt.preventDefault();

    if (todoInput.trim() === '') {
      setIsInvalid(!initialState.isInvalid);
      setTodoInput(initialState.todoInput);
    } else {
      props.addTodo(todoInput);
      setIsInvalid(initialState.isInvalid);
      setTodoInput(initialState.todoInput);
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <TextInput
        name="newTodo"
        value={todoInput}
        styleName={styles}
        placeholder={placeholder}
        onChange={handleChange}
        onBlur={handleBlur}
      />
    </form>
  );
}

TodoForm.propTypes = {
  addTodo: PropTypes.func.isRequired
};
