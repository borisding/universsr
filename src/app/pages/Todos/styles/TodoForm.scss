@import '@common/styles/variables';

.todo-input {
  border: 2px solid #ccc;
  font-weight: 400;
  margin-bottom: $base-margin;
  padding: $base-padding / 2;
}

.error-input {
  border-color: #ec0225;
  outline: none;
}
