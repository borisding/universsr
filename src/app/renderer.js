import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
