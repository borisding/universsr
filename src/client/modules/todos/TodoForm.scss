@import '../../common/styles/variables';

.todoInput {
  font-size: $base-font-size + 4;
  padding: ($base-padding / 3);
}
