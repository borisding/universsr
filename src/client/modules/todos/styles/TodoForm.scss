@import '~@styles/variables';

.todo-input {
  border: 2px solid #ccc;
  font-size: $base-font-size + 2;
  margin-bottom: $base-margin;
  padding: $base-padding / 2;
}

.error-input {
  border-color: #ec0225;
  outline: none;
}
