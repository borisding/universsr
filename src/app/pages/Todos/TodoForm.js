import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { TextInput } from '@common/components';
import './styles/TodoForm.scss';

const initialState = {
  todoInput: '',
  isInvalid: false
};

export default class TodoForm extends Component {
  static propTypes = {
    addTodo: PropTypes.func.isRequired
  };

  constructor() {
    super();
    this.state = initialState;
  }

  handleChange = evt => {
    this.setState({
      todoInput: evt.target.value,
      isInvalid: initialState.isInvalid
    });
  };

  handleBlur = evt => {
    if (this.state.isInvalid) {
      this.setState({
        todoInput: evt.target.value,
        isInvalid: !this.state.isInvalid
      });
    }
  };

  handleSubmit = evt => {
    evt.preventDefault();
    const { todoInput } = this.state;

    if (todoInput.trim() === '') {
      return this.setState({
        todoInput,
        isInvalid: !initialState.isInvalid
      });
    }

    return this.props
      .addTodo(todoInput)
      .then(() => this.setState(initialState));
  };

  render() {
    const { todoInput, isInvalid } = this.state;
    const styles = classNames('todo-input', { 'error-input': !!isInvalid });
    const placeholder = isInvalid ? 'Todo is required!' : 'Enter new todo...';

    return (
      <form onSubmit={this.handleSubmit}>
        <TextInput
          name="newTodo"
          value={todoInput}
          styleName={styles}
          placeholder={placeholder}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </form>
    );
  }
}
