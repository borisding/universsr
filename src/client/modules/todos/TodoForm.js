import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TextInput from '@common/components/TextInput';
import './styles/TodoForm.scss';

const initialState = {
  todoInput: '',
  isInvalid: false
};

export default class TodoForm extends Component {
  static propTypes = {
    addTodo: PropTypes.func.isRequired
  };

  constructor() {
    super();
    this.state = initialState;
    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(evt) {
    this.setState({
      todoInput: evt.target.value,
      isInvalid: initialState.isInvalid
    });
  }

  handleBlur(evt) {
    if (this.state.isInvalid) {
      this.setState({
        todoInput: evt.target.value,
        isInvalid: !this.state.isInvalid
      });
    }
  }

  handleSubmit(evt) {
    evt.preventDefault();
    const { todoInput } = this.state;

    if (todoInput.trim() === '') {
      return this.setState({
        todoInput,
        isInvalid: !initialState.isInvalid
      });
    }

    return this.props
      .addTodo(todoInput)
      .then(() => this.setState(initialState));
  }

  render() {
    const { todoInput, isInvalid } = this.state;

    return (
      <form onSubmit={this.handleSubmit}>
        <TextInput
          name="newTodo"
          value={todoInput}
          styleName={'todo-input' + (isInvalid ? ' error-input' : '')}
          placeholder={isInvalid ? 'Todo is required!' : 'Enter new todo...'}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </form>
    );
  }
}
