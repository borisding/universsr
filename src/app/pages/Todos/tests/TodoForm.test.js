import React from 'react';
import renderer from 'react-test-renderer';
import TodoForm from '../TodoForm';

describe('TodoForm', () => {
  it('renders correctly', () => {
    const props = { addTodo: jest.fn() };
    const tree = renderer.create(<TodoForm {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
