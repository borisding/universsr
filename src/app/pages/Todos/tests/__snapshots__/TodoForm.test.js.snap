// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    className="todo-input"
    name="newTodo"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Enter new todo..."
    type="text"
    value=""
  />
</form>
`;
